### DISCUSSION TITLE
setting revolute joint position?

### DISCUSSION THREAD

**Discussion by mattdm** [avianphysics/avian#181] [General]
Rapier3d joints have "motors", which ... [aren't working for me like I think they ought to](https://discord.com/channels/691052431525675048/1163281784461607052), so maybe I'm just really confused overall. But what I want is: a revolute joint which can swivel under normal influences subject to its limits but returns (with configurable torque and dampening) to a set point. I'd like to be able to change that set point dynamically -- causing a turret to swivel, or a head to turn, or a door to open or close.

However, xpbd does not have motors. There is `position_lagrange` and `align_lagrange`, but even after reading the theory section I'm not really clear on what these are or why I would want to set them... and anyway this doesn't seem like _it_.

I guess I could temporarily reduce the angle limits, and then reset them? That doesn't seem right, though. What am I missing?

**Comment by Jondolf** [comment]
I don't think you can make proper motors right now, it's something that will be added in the future. You *might* be able to cheat it by changing the angle limits, but I doubt it would be very robust or easy to control

**Comment by mattdm** [comment]
I am super-naïve here, having little real physics background _or_ experience with other game engines. So, feel free to back away slowly if this is all completely off-base.

I see that RevoluteJoint has [`align_orientation`](https://docs.rs/bevy_xpbd_3d/latest/bevy_xpbd_3d/constraints/joints/trait.Joint.html#method.align_orientation) (well, I guess all joints do, but... starting there)... until there _are_ motors, to make a system that has the same effect, would that be the function to use? That takes a `delta_q` ... is that angle to try to rotate along each axis? If so, what is this vector aligned to (hopefully not _world_ x,y,z axes!)? Or, is it a torque vector? Or something else entirely?

This function also takes "lagrange" and "compliance" -- aren't these already properties of the joint itself, though? I see that it _returns_ torque — I'm having trouble thinking of things from this (metaphorical) perspective — I'm thinking of a joint as a thing which _applies force to the attached bodies to cause them to align in a certain way_, and maybe that's all wrong.

Maybe a motor shouldn't be part of a joint at all, but an entirely separate thing which:

* attaches to a rigid body
* applies force or torque to any other rigid body in contact at that point 
    * taking into account friction between those bodies
        * possibly with a parameter to "lock" and always use an infinite coefficient of friction, so there is no slip
* and then the bodies would follow the physics model, including being held by any joints

For convenience, it would be of course nice to have a way to perfectly align a joint and motor. 

I'm thinking mostly of rotations, where I think the "motor" would be a point (although, interesting if it could be a circle, cylinder, or sphere, actually...). For linear motors, I would expect a line in 2D or rectangle (or other shape?) in 3D. I'm not sure there's a real-world equivalent, but for a game, this would make it easy to make those speed-up ramps in Mario Kart, or moving sidewalks without actually moving a platform.

**Comment by mattdm** [comment]
Leaving aside motor ideas _in_ xbpd for a second... if `align_orientation` isn't the right approach, my other idea is to make a system which:

* looks for entities with `RigidBody` and some component like "Servomotor"
* reads an angle from that
* and either also an entity id _or_ finds a joint and gets the other RigidBody entity from that
* gets the rotational angle between the RigidBody with the Servomotor and the other RigidBody
* if the difference is bigger than some configured precision, and an ExternalForce with torque in the right direction to the other RigidBody
* possibly scale the torque to the delta? I dunno. I will have to experiment.


This seems more straightforward to me, but less... integrated. I'd love to hear opinions on if this is close to sensible. :)
*

**Comment by RunnersNum40** [comment]
I see that motors aren't implemented yet, but in the interim does anyone have a good way to achieve the same functionality?

I'm curious if there's a possible approach of applying torques and forces to the entities themselves.
